File: dev/producer.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"os"
	"strconv"
	"time"
)

func sendMessagesToBroker(pdType int, matrixSize int) {
	brokerURL := "http://consumer:8080"
	endURL := "http://consumer:8080/end"
	i := 0
	for i < matrixSize*matrixSize {
		if i%1000 == 0 {
			fmt.Printf("Sending message %d...\n", i)
		}
		data := map[string]interface{}{
			"message_type":    pdType,
			"message_content": rand.Intn(1000) + 1,
		}

		// Convert data to JSON
		jsonData, err := json.Marshal(data)
		if err != nil {
			fmt.Printf("Error encoding JSON: %v\n", err)
			continue
		}

		// Send POST request with JSON data
		response, err := http.Post(brokerURL, "application/json", bytes.NewBuffer(jsonData))
		if err != nil {
			fmt.Printf("Error sending message: %v\n", err)
			time.Sleep(1 * time.Second)
			continue
		}
		if response.StatusCode == -520 {
			continue
		}
		i++
	}

	data := map[string]interface{}{
		"message_type":    pdType,
		"message_content": -1,
	}

	// Convert data to JSON
	jsonData, err := json.Marshal(data)
	if err != nil {
		fmt.Printf("Error encoding JSON: %v\n", err)
		return
	}

	// Send POST request with JSON data
	_, err = http.Post(endURL, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		fmt.Printf("Error sending message: %v\n", err)
		time.Sleep(1 * time.Second)
	}
}

func main() {
	if len(os.Args) > 2 {
		arg1, err := strconv.Atoi(os.Args[1])
		if err != nil {
			fmt.Println("Invalid argument provided. Should be a producer type: 1 or 2")
			return
		}
		arg2, err := strconv.Atoi(os.Args[2])
		if err != nil {
			fmt.Println("Invalid argument provided. Should be a producer type: 1 or 2")
			return
		}
		sendMessagesToBroker(arg1, arg2)
	} else {
		fmt.Println("No arguments provided. Should be provided producer type: 1 or 2 and number of messages to send")
	}
}

File: producer1/producer.py
import random
import requests
import time
import sys
import os

MSize = int(os.getenv('MSize', 4))
def send_messages_to_broker(PDtype):
    broker_url = 'http://consumer:8080'
    i = 0
    while i < MSize ** 2:
        if i % 100 == 0:
            print(f"Sending message {i}...")
        try:
            data = {
                'message_type': PDtype,
                'message_content': random.randint(1, 1000),
            }
            response = requests.post(broker_url, json=data)

            if response.status_code != 200:
                print(f"Received invalid response: {response.status_code}, {response.text}")
                continue

            while response.text == "-520":
                print("Received -520, retrying...")
                time.sleep(1)
                response = requests.post(broker_url, json=data)

            i += 1
        except requests.exceptions.RequestException as E:
            print(f"Error sending message: {E}")
            time.sleep(1)

    try:
        data = {
            'message_type': PDtype,
            'message_content': -1,
        }
        response = requests.post(broker_url + "/end", json=data)
        if response.status_code == 200:
            print("End signal sent successfully.")
        else:
            print(f"Error sending end signal: {response.status_code}, {response.text}")
    except requests.exceptions.RequestException as E:
        print(f"Error sending end message: {E}")
        time.sleep(1)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        try:
            arg = int(sys.argv[1])
            if arg not in [1, 2]:
                print("Invalid producer type. Should be 1 or 2.")
                sys.exit(1)
            send_messages_to_broker(arg)
        except ValueError:
            print("Invalid argument. Producer type should be an integer (1 or 2).")
            sys.exit(1)
    else:
        print("No arguments provided. Should be provided producer type: 1 or 2")
